### PR TITLE
Allow changing number of replicas of searchable snapshot index

### PR DESCRIPTION
**Pull request by bugmakerrrrrr**
### Description
#5907 uses a whitelist to control which settings of searchable snapshot can be changed, but `index.number_of_shards` is not on the list. I think it is necessary to allow user to change the number of replicas based on their cost or performance needs.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by bugmakerrrrrr** [comment]
@andrross what do you think about this idea?

File: server/src/main/java/org/opensearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java [hunk 1]
```diff
@@ -145,10 +147,10 @@ protected ClusterBlockException checkBlock(UpdateSettingsRequest request, Cluste
             }
         }
 
+        final String[] requestIndexNames = Arrays.stream(requestIndices).map(Index::getName).toArray(String[]::new);
         return allowSearchableSnapshotSettingsUpdate
             ? null
-            : state.blocks()
-                .indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, indexNameExpressionResolver.concreteIndexNames(state, request));
+            : state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, requestIndexNames);
```
Comment (andrross) [inline server/src/main/java/org/opensearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:153]: Why is this change here? How is this related to adding support for `number_of_replicas`?

**Review by andrross — COMMENTED** [review]
Thanks @bugmakerrrrrr! This is definitely a gap as configuring replicas is something that should be allowed.

File: server/src/main/java/org/opensearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java (hunk 1 above)
Comment (bugmakerrrrrr) [reply server/src/main/java/org/opensearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:153]: I just think we can reuse the first result to avoid computing request indices twice. This has nothing to do with adding support for `number_of_replicas`

**Comment by andrross** [comment]
Thanks @bugmakerrrrrr. This change looks good. Can you please add an entry to the changelog under `[Unreleased 2.x]`?

**Comment by bugmakerrrrrr** [comment]
@andrross Done. Btw, I find that allow changing allocation filter settings (`index.routing.allocation.include/exclude/require.*`) of searchable snapshot may also be necessary. Can I add support for it in this PR? or should I open a new PR?

**Comment by andrross** [comment]
> @andrross Done. Btw, I find that allow changing allocation filter settings (`index.routing.allocation.include/exclude/require.*`) of searchable snapshot may also be necessary. Can I add support for it in this PR? or should I open a new PR?

Let's do that in a separate PR. Feel free to open an issue to track it as well.
